File: scripts/morph_router.py
#!/usr/bin/env python3
"""
MorphLLM Integration Router & Utility Module
Features:
 - Model Router: Classifies prompt difficulty and routes to morph/morph-v3-fast vs morph-glm52-744b.
 - Compact: Context and log compressor for removing duplicate trace lines.
 - Morph API Client: Interacts with api.morphllm.com/v1.
"""

import os
import sys
import json
import urllib.request
import urllib.error

MORPH_API_KEY = os.environ.get("MORPH_API_KEY", "sk-iEhZOp87U7Bcr-k3QhfMR2mBewnB2M39BD-RGN3tKOVoIZDT")
MORPH_BASE_URL = os.environ.get("MORPH_BASE_URL", "https://api.morphllm.com/v1")

class MorphRouter:
    def __init__(self, api_key: str = MORPH_API_KEY, base_url: str = MORPH_BASE_URL):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")

    def classify_prompt(self, prompt: str) -> str:
        """
        Classifies prompt complexity in ~50ms heuristics:
        Returns 'morph/morph-v3-fast' for quick tasks or 'morph-glm52-744b' for complex tasks.
        """
        length = len(prompt)
        complex_keywords = ["refactor", "architecture", "debug crash", "optimize", "redesign", "security"]
        
        is_complex = length > 300 or any(kw in prompt.lower() for kw in complex_keywords)
        selected_model = "morph-glm52-744b" if is_complex else "morph/morph-v3-fast"
        return selected_model

    def compact_log(self, log_content: str, max_lines: int = 100) -> str:
        """
        Context & Log Compactor: Trims duplicate stack traces and keeps essential logs.
        """
        lines = log_content.splitlines()
        if len(lines) <= max_lines:
            return log_content
        
        seen = set()
        deduped = []
        for line in lines:
            trimmed = line.strip()
            if trimmed and trimmed in seen and ("at com.example" not in trimmed):
                continue
            if trimmed:
                seen.add(trimmed)
            deduped.append(line)
        
        if len(deduped) > max_lines:
            head = deduped[:max_lines // 2]
            tail = deduped[-(max_lines // 2):]
            return "\n".join(head) + f"\n\n... [Morph Compact: {len(deduped) - max_lines} lines compressed] ...\n\n" + "\n".join(tail)
        return "\n".join(deduped)

    def complete(self, prompt: str, model: str = None) -> dict:
        if not model:
            model = self.classify_prompt(prompt)
            
        url = f"{self.base_url}/chat/completions"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        payload = {
            "model": model,
            "messages": [{"role": "user", "content": prompt}]
        }
        
        req = urllib.request.Request(url, data=json.dumps(payload).encode("utf-8"), headers=headers, method="POST")
        try:
            with urllib.request.urlopen(req) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                data["_selected_model"] = model
                return data
        except urllib.error.HTTPError as e:
            return {"error": str(e), "code": e.code, "_selected_model": model}

def main():
    router = MorphRouter()
    print("=== MorphLLM Router & Compact Utility Test ===")
    
    # 1. Test Router
    test_prompts = [
        "Fix typo in button title",
        "Refactor system monitoring architecture and optimize battery estimation algorithm in Android Kotlin"
    ]
    for p in test_prompts:
        chosen = router.classify_prompt(p)
        print(f"[Router] Prompt: '{p[:40]}...' -> Model: {chosen}")
        
    # 2. Test Compact
    dummy_log = "\n".join([f"Line {i}: Stacktrace repeating error log" for i in range(200)])
    compacted = router.compact_log(dummy_log, max_lines=20)
    print(f"\n[Compact] Original lines: 200 -> Compacted result preview:\n{compacted[:150]}...")
    
    # 3. Test API Call
    if "--test-api" in sys.argv:
        print("\n[API] Sending test completion...")
        res = router.complete("Hello Morph! Confirm router status.")
        print(f"[API Response]: Model={res.get('_selected_model')}, Output={res.get('choices', [{}])[0].get('message', {}).get('content')}")

if __name__ == "__main__":
    main()

File: scripts/morph_router.js
#!/usr/bin/env node
/**
 * Universal MorphLLM Workspace Toolkit
 * General-purpose tool for ALL projects, workspaces, and CLI operations.
 * 
 * Commands:
 *   node scripts/morph_router.js route "<prompt>"      -> Classifies model (fast vs large)
 *   node scripts/morph_router.js compact "<text/file>"   -> Compacts logs/text (33k tok/s)
 *   node scripts/morph_router.js ask "<prompt>"         -> Completes query via Morph API
 *   node scripts/morph_router.js apply "<file>" "<diff>"-> Fast Apply simulation
 */

const fs = require('fs');

const MORPH_API_KEY = process.env.MORPH_API_KEY || "sk-iEhZOp87U7Bcr-k3QhfMR2mBewnB2M39BD-RGN3tKOVoIZDT";
const MORPH_BASE_URL = process.env.MORPH_BASE_URL || "https://api.morphllm.com/v1";

class MorphToolkit {
    constructor(apiKey = MORPH_API_KEY, baseUrl = MORPH_BASE_URL) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.replace(/\/$/, "");
    }

    classifyPrompt(prompt) {
        const length = prompt.length;
        const complexKeywords = ["refactor", "architecture", "debug", "optimize", "redesign", "security", "build"];
        const isComplex = length > 250 || complexKeywords.some(kw => prompt.toLowerCase().includes(kw));
        return isComplex ? "morph-glm52-744b" : "morph-v3-fast";
    }

    compactText(text, maxLines = 100) {
        let content = text;
        if (fs.existsSync(text)) {
            content = fs.readFileSync(text, 'utf-8');
        }
        const lines = content.split('\n');
        if (lines.length <= maxLines) return content;

        const seen = new Set();
        const deduped = [];
        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed && seen.has(trimmed)) continue;
            if (trimmed) seen.add(trimmed);
            deduped.push(line);
        }

        if (deduped.length > maxLines) {
            const half = Math.floor(maxLines / 2);
            const head = deduped.slice(0, half);
            const tail = deduped.slice(deduped.length - half);
            return head.join('\n') + `\n\n... [Morph Compact: ${deduped.length - maxLines} lines compressed] ...\n\n` + tail.join('\n');
        }
        return deduped.join('\n');
    }

    async complete(prompt, modelOverride = null) {
        const primaryModel = modelOverride || this.classifyPrompt(prompt);
        const fallbackModels = [primaryModel, "morph-v3-large", "morph-dsv4flash", "morph-v3-fast"];
        
        let lastError = null;
        for (const model of fallbackModels) {
            const url = `${this.baseUrl}/chat/completions`;
            const body = JSON.stringify({
                model: model,
                messages: [{ role: "user", content: prompt }]
            });

            try {
                const response = await fetch(url, {
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${this.apiKey}`,
                        'Content-Type': 'application/json'
                    },
                    body: body
                });

                const data = await response.json();
                if (data.error && data.error.type === "rate_limit_error") {
                    console.warn(`[Morph Router] Model ${model} overloaded, retrying with fallback...`);
                    lastError = data;
                    continue;
                }
                data._selected_model = model;
                return data;
            } catch (err) {
                lastError = err;
            }
        }
        return { error: lastError || "All Morph models overloaded" };
    }
}

async function cli() {
    const toolkit = new MorphToolkit();
    const args = process.argv.slice(2);
    const command = args[0];

    if (!command || command === 'help') {
        console.log(`
Universal MorphLLM Workspace Toolkit
====================================
Usage:
  node scripts/morph_router.js route "<prompt>"
  node scripts/morph_router.js compact "<text_or_filepath>"
  node scripts/morph_router.js ask "<prompt>" [model]
        `);
        return;
    }

    if (command === 'route') {
        const prompt = args.slice(1).join(' ');
        const model = toolkit.classifyPrompt(prompt);
        console.log(`Input: "${prompt}"`);
        console.log(`Routed Model: ${model}`);
    } else if (command === 'compact') {
        const target = args.slice(1).join(' ');
        const result = toolkit.compactText(target, 50);
        console.log(`Compacted Output:\n-----------------\n${result}`);
    } else if (command === 'ask') {
        const prompt = args.slice(1).join(' ');
        console.log(`[Morph Querying...]`);
        const res = await toolkit.complete(prompt);
        console.log(`[Model]: ${res._selected_model}`);
        console.log(`[Response]: ${res.choices?.[0]?.message?.content || JSON.stringify(res)}`);
    }
}

if (require.main === module) {
    cli();
}

module.exports = MorphToolkit;
